Batteries/Data/Fin/Lemmas.lean

```diff
@@ -126,31 +134,49 @@ theorem find?_succ {p : Fin (n+1) → Bool} :
   simp only [find?, findSome?_succ, Option.guard]
   split <;> simp [map_findSome?, Function.comp_def, Option.guard]
 
-theorem eq_true_of_find?_eq_some {p : Fin n → Bool} (h : find? p = some i) : p i = true := by
-  match exists_of_findSome?_eq_some h with
-  | ⟨i, hi⟩ =>
-    simp only [Option.guard] at hi
-    split at hi
-    · cases hi; assumption
-    · contradiction
-
-theorem eq_false_of_find?_eq_none {p : Fin n → Bool} (h : find? p = none) (i) : p i = false := by
-  have hi := eq_none_of_findSome?_eq_none h i
-  simp only [Option.guard] at hi
-  split at hi
-  · contradiction
-  · simp [*]
-
-theorem find?_isSome_iff {p : Fin n → Bool} : (find? p).isSome ↔ ∃ i, p i := by
+theorem find?_eq_some_iff {p : Fin n → Bool} :
+    find? p = some i ↔ p i ∧ ∀ j, j < i → ¬ p j := by
+  simp [find?, and_assoc]
+
+theorem find?_isSome_iff {p : Fin n → Bool} :
+    (find? p).isSome ↔ ∃ i, p i ∧ ∀ j < i, ¬ p j := by
+  simp [find?]
+
+theorem find?_eq_none_iff {p : Fin n → Bool} : find? p = none ↔ ∀ i, ¬ p i := by
   simp [find?]
 
 theorem find?_isNone_iff {p : Fin n → Bool} : (find? p).isNone ↔ ∀ i, ¬ p i := by
   simp [find?]
 
-proof_wanted find?_eq_some_iff {p : Fin n → Bool} : find? p = some i ↔ p i ∧ ∀ j, j < i → ¬ p j
+theorem eq_true_of_find?_eq_some {p : Fin n → Bool} (h : find? p = some i) :
+    p i ∧ ∀ j < i, ¬ p j := by
+  rwa [find?_eq_some_iff] at h
 
-theorem find?_eq_none_iff {p : Fin n → Bool} : find? p = none ↔ ∀ i, ¬ p i := by
-  rw [← find?_isNone_iff, Option.isNone_iff_eq_none]
+theorem eq_false_of_find?_eq_none {p : Fin n → Bool} (h : find? p = none) (i) :
+    ¬ p i := by
+  rw [find?_eq_none_iff] at h
+  grind
+
+theorem exists_of_find?_isSome {p : Fin n → Bool} (h : (find? p).isSome) :
+    ∃ i, p i ∧ ∀ j < i, ¬ p j := by
+  rwa [find?_isSome_iff] at h
+
+theorem exists_of_find?_isSome' {p : Fin n → Bool} (h : (find? p).isSome) :
+    ∃ i, p i := by
+  have H := exists_of_find?_isSome h
+  grind
+
+theorem isNone_of_find?_isNone (h : (find? p).isNone) : ¬ p i := by
```

Comment:
```suggestion
theorem ne_true_of_isNone_find? (h : (find? p).isNone) : ¬ p i := by
```
Add `eq_false_of_isNone_find?` too?

Reply:
As per Zulip, I think we should consider pos/neg. I think we should probably avoid eq_false and treat p like a predicate.